text/0141-linking-traces.md

```diff
@@ -0,0 +1,367 @@
+Importantly, we must never require span links to be defined. They are completely optional.
+
+TODO: This section lacks a lot of details we still need to clarify with the ingest team
+
+### Storage
+
+We need to adapt our events table to support storing span links in our current event storage. Furthermore, we need to take span links into account for the EAP storage architecture.
+
+TODO: Evaluate if possible:
+
+The span links need to be stored in a way that they can be queried. For example: Given a transaction `t`, list transactions that have span links to  `t`.
+This would enable use cases where we could show "next" traces, instead of previous traces. 
+
+TODO: This section lacks a lot of details we still need to clarify with the SnS team
+
+
+## [Alternative] Previous Trace Id
+
+A slightly simpler but less powerful approach to span links would be to specifically store a "previous trace id" attribute on the current root span and serialize it as such in the event envelope item.
+Setting such a previous trace id could happen on `startSpan()` calls or by calling `span.setAttribute()` at any time. 
+
+While this might seem simpler at first glance, it has a some drawbacks:
+- Spans can only have a 1:1 relationship, either with a `traceId` or a `traceId-spanId` value. Span links support a 1:* relationship, where one span can link to multiple other spans in- or outside its trace
+- Span links are an established concept in OpenTelemetry. With Sentry increasingly adapting Otel standards, as well as creating an OTLP endpoint in the future, we should embrace Otel-suggested solutions rather than building our own.
+- Attributes are not indexed, meaning queries like getting the "next" instead of previous traces are likely impossible
+- POtel SDKs, as well as the browser JS SDKs, already expose span link APIs but they currently no-op and are not sent to Sentry. This is currently very opaque behavior and potentially confusing to users familiar with Otel APIs concepts.
+
+
+# Drawbacks
+
+- Implementing span links requires a lot of of changes across SDKs, ingest, SnS. Also there are product changes required to make use of links. 
+
+# Answered Questions
+
+- Can we do this with our current event storage? Are we blocked on EAP?
+  Yes. We will certainly need to adapt our current event storage but it is possible to do this without being blocked on EAP. We'll separately also need to adjust the planned span schema for EAP but this is not a blocker.
+
+
+- How would we handle errors outside of an active span?
+  Errors happening while no span is active, should (as currently is the case in JS Browser SDKs) belong to the last created trace. Therefore, they will store all required fields in their `trace` context.
+
+# Unresolved Questions
+
+- Should we (try to) keep sampling decisions consistent for subsequent (frontend) traces? IOW: If we sample the initial trace, should all subsequent traces be force-sampled as well? 
```

Comment:
This seems like the key question for this proposal to me - otherwise, the odds of having links are very low, especially an uninterrupted chain of them, given client and server sampling. I'm not sure if the feature is useful if the traces are still sampled independently.

Reply:
I agree that having a consistent sampling decision across linked traces would make sense. There are some caveats though which we need to consider and deal with. I will discuss this with Ingest folks but will update the RFC ASAP.

* Different sampling mechanisms
  * Assuming, we have a unifom `tracesSampleRate`, quota usage should be consistent, regardless if we continue the sampling decision across traces or sample individually. So the advantage here would obviously be a more consistent user journey vs. only partial signals. 
  * However, if users specify a `tracesSampler` function, they could increase or decrease the sampling of certain routes or spans which means the number of events (i.e. quota usage) could drastically change as opposed to individual sampling. 
  * The `tracesSampler` function can overrule the sampling decision of upstream SDKs in case of continuing a remote trace. Maybe we can leverage the same mechanism to overrule sampling for previous traces? This is TBD though because there are discussions about no longer permitting overruling the remote sampling decision.
* Multi Page Applications
  * Let's assume we have a multi page app where each page navigation triggers a full page reload cycle. This means that for every page navigation, we'd have a new trace **starting on the server** that renders and returns the page. The problem here is that we cannot propagate the sampling decision of the previous trace to the next server-side request. That is, as long as we don't set a cookie (which is a line I don't think we want to cross). So there will be inconsistent sampling in MPAs.
  * We can still connect the individual page load spans on the client side and thereby connect the two traces but we can't enforce consistent sampling
  * We _could_ let users define their own cookie with some mechanism to work around this but I'd argue this is out of scope. 
  * So if we go with consistent sampling across traces, we'll need to accept this limitation.

Reply:
I addressed the sampling question in the "Answered questions" section.